avro2tf/src/main/scala/com/linkedin/avro2tf/parsers/TensorizeInConfigParser.scala

```diff
@@ -1,35 +1,93 @@
-import org.json4s.DefaultFormats
-import org.json4s.jackson.JsonMethods.parse
+import com.typesafe.config.{Config, ConfigFactory}
+import io.circe
+import io.circe.Errors
+import io.circe.generic.auto._
+import io.circe.config.syntax._
+
+import collection.JavaConverters._
+import scala.util.Try
 
 /**
- * Parser file for TensorizeIn configuration
- *
- */
+  * Parser file for TensorizeIn configuration
+  *
+  */
 object TensorizeInConfigParser {
 
   /**
    * Get the TensorizeIn configuration with sanity check
    *
-   * @param jsonString JSON format of TensorizeIn configuration
+   * @param configString HOCON format of TensorizeIn configuration
    * @return TensorizeIn configuration
    */
-  def getTensorizeInConfiguration(jsonString: String): TensorizeInConfiguration = {
+  def getTensorizeInConfiguration(configString: String): TensorizeInConfiguration = {
+
+    val rawConfig = ConfigFactory.parseString(configString)
+    val featureConfig = rawConfig.getConfigList("features").asScala.map(parseFeature)
+    val labelConfig = Try(rawConfig.getConfigList("labels")).map(_.asScala.map(parseFeature)).toOption
 
-    // Define implicit JSON4S default format
-    implicit val formats: DefaultFormats.type = DefaultFormats
+    // check that we parsed the features correctly
+    val errorList = NonEmptyList.fromList((featureConfig ++ labelConfig.getOrElse(List()))
+      .filter(_.isLeft).map(_.left.get).toList)
+    errorList.foreach(es => throw Errors(es).fillInStackTrace())
 
-    // Use JSON4S to parse and extract TensorizeIn configuration
-    val config = parse(jsonString).extract[TensorizeInConfiguration]
+    // since we didn't throw any errors, we know everything is a Right
+    val features = featureConfig.map(_.right.get)
+    val labels = labelConfig.map(_.map(_.right.get))
 
     TensorizeInConfiguration(
-      features = sanityCheck(config.features),
-      labels = Some(sanityCheck(config.labels.getOrElse(Seq.empty)))
+      features = sanityCheck(features),
+      labels = Some(sanityCheck(labels.getOrElse(Seq.empty)))
     )
   }
 
+  /** Parse the feature configs.
+    *
+    * @param featureConfig - the [[Config]] object for an individual [[Feature]]
+    * @return [[Either]] of a [[circe.Error]] or a [[Feature]]
+    */
+  private def parseFeature(featureConfig: Config): Either[circe.Error, Feature] = {
+    val outputTensorInfo: Either[circe.Error, OutputTensorInfo] =
+      featureConfig.getConfig("outputTensorInfo").as[OutputTensorInfo]
+
+    val inputFeatureInfo = if (featureConfig.hasPath("inputFeatureInfo"))
+      Some(parseInputFeatureInfo(featureConfig.getConfig("inputFeatureInfo")))
+    else None
+
+    outputTensorInfo.fold(e => Left(e), oti => Right(Feature(inputFeatureInfo, oti)))
+  }
+
+  /**
+    *
+    * @param inputFeatureInfoConfig
+    * @return
+    */
+  private def parseInputFeatureInfo(inputFeatureInfoConfig: Config): InputFeatureInfo = {
```

Comment:
Could you provide some comment here why we need a separate parser for `InputFeatureInfo`. is it mainly because we want to convert some fields to `Option` type? Then how about the `shape` field in `OutputTensorInfo`? do we need a special handling here? Thanks.

Reply:
It's because the transform config has type `Map[String, Map[String, Any]]`--none of the Scala libraries for typesafe config seem to support decoding to `Map[String, Any]`. I think the right solution is to have `HashInfo` and (a new)`TokenizationInfo` extend the same sealed trait, call it `TransformType`. Then the transform config becomes `Map[String, Map[String, TransformType]]` and you should be able to decode, complete with nice exceptions.

I'm happy to make this change, but it involves touching other parts of the codebase beyond just the config, so I wanted to talk about it first.

Reply:
Actually, looking at this some more, the transform config should probably be a case class with two optional fields `hashInfo` and `tokenization`. That gets rid of the sealed trait and enforces the field names in the config as it's parsed.